### ISSUE TITLE
Python-bound C++ frontend modules don't handle attributes well

### ISSUE THREAD

**Issue by goldsborough** [pytorch/pytorch#16375]
When we bind C++ frontend modules into Python using `torch::python::bind_module` and load them in Python, we wrap the underlying C++ class with a [wrapper class](https://github.com/pytorch/pytorch/blob/master/torch/nn/cpp.py#L49) that derives from `torch.nn.Module` and copies all methods of the underlying class onto the wrapper class (I can elaborate on design reasons for this).

The issue is that this does not support attributes. As in, if you bind an attribute in C++ using `def_readwrite("x", &MyClass::x)`, and try setting that attribute on the Python wrapper, it will not propagate. This should be handled by intelligently overriding `__getattr__` and `__setattr__`.

CC @yf225

**Comment by huntzhan** [comment]
Hi @goldsborough, I'm playing around Python-bound C++ frontend modules recently and notice that there're some gaps in usability. Not sure if this the right place to discuss.

When bound to Python, the C++ module will not be treated the same as `torch.nn.Module`. As a consequence, I need to [manually bind the parameters](https://github.com/cnt-dev/pytorch-fast-elmo/blob/d78b64e2f692068bb3ba6d76a841e47f3b65d7f3/pytorch_fast_elmo/integrate.py#L20-L42) of C++ module to the parent module to support methods like `cuda(device)` or `cpu()`. It will be must better if Pytorch will handle this automatically.

Also, I notice that the Python-bound methods do not have the same signature as the corresponding Pytorch implementation. Take `cuda` as an example, the [C++ implementation](https://github.com/pytorch/pytorch/blob/29827e19711ce5a73adcd54f4b0ec694d0b21e79/torch/csrc/api/include/torch/python.h#L183) doesn't accept a `device` parameter while the Pytorch version does accept. I have to [manually add a cuda binding](https://github.com/cnt-dev/pytorch-fast-elmo/blob/d78b64e2f692068bb3ba6d76a841e47f3b65d7f3/extension/bind.cc#L22-L26) to work around.

The C++ frontend is great and thanks for your work!

**Comment by goldsborough** [comment]
How did you bind your C++ module? If you use `torch::python::bind_module` it will be a `torch.nn.Module` in Python. See [the tests](https://github.com/pytorch/pytorch/blob/master/test/cpp_extensions/cpp_frontend_extension.cpp#L47) for an example.

**Comment by huntzhan** [comment]
@goldsborough Yeah I've used the `torch::python::bind_module` to bind ([my code](https://github.com/cnt-dev/pytorch-fast-elmo/blob/d78b64e2f692068bb3ba6d76a841e47f3b65d7f3/extension/bind.cc#L37-L38)).

```
In [20]: from _pytorch_fast_elmo import ScalarMix                                                   

In [21]: scm = ScalarMix(3)                                                                         

In [22]: type(scm)                                                                                  
Out[22]: _pytorch_fast_elmo.ScalarMix

In [23]: isinstance(scm, torch.nn.Module)                                                           
Out[23]: False
```

This was tested in torch 1.0.0.

**Comment by bumi001** [comment]
@goldsborough, you mentioned back in 2019 that you can elaborate the design reasons for wraping the underlying C++ class with a [wrapper class](https://github.com/pytorch/pytorch/blob/master/torch/nn/cpp.py#L49) that derives from torch.nn.Module. I am wondering if the underlying class is already derived from torch::nn::module, why not pass that to detail::bind_cpp_module_wrapper directly?
